doc/specs/#605 - Search/spec.md

```diff
@@ -0,0 +1,91 @@
+2. The user search in a XAML AutoSuggestBox on the top right corner of the Terminal window. 
+3. We can have multiple search methods. The simplest one is text exact match. Other match methods include case-sensitive exact match and regex match. In the first phrase, we will focus on case sensitive/insensitive text exact match. 
+4. If currently there is no active selection, the search starts from the line that the cursor is on. If there is an active selection, we start from the previous or the next text of the selected text. We automatically go around if we reach the start point of the search. 
+5. The search dialog should not block terminal's view. 
+
+Conhost already has a module for search. It realizes case sensitive or insensitive exact text match search, and it provides methods to select the found word. However, we want to make search as a shared component between Terminal and Console host. Now search module is part of Conhost, and its dependencies include BufferOut and some other types in ConHost such as SCREEN_INFORMATION. In order to make Search a shared component, we need to remove its dependency on ConHost types. BufferOut is already a shared component, but we need to make sure there is no other Conhost dependency.
+
+Search process:
+Search is performed on a XAML AutoSuggestBox. Once the user click the "Find" icon, we start from the cursor (or the current selection), and try to find the exact text in the text buffer. The nearest searched one will be selected. And we set the search start point to the selected text. The next time "Find" button is clicked, the search will start before or after the previous searched text.
+The user can choose to search up or down by choosing up arrow or down arrow buttons.
+The user can choose to do case sensitive or insensitive match by clicking a button.
+If the user click the "X" button, the search stopped and the search box disappears. In phrase one we do not store any state, but in the future we can consider storing the search history. 
+
+## UI/UX Design
+
+![Sol Design](images/AutoSuggestBox.png)
+
+Above is the XAML AutoSuggestBox. We also need to add at least two more buttons, one for case sensitive/insensitive switch, and one for search direction switch. We can consider adding the buttons on the right side of the AutoSuggestBox.
+
+The search box should be on the top right corner of the Terminal window. We need to avoid it blocking too much screen contents. 
+
+## Capabilities
+
+1. The user can search exact matched text in the text buffer of the Terminal Screen. 
+2. The user can choose to search case sensitively and insensitively. 
+3. The user can search up or down. 
+4. Found text will be selected. 
+5. The search will start from the active selected text (inclusive) if there is one, or where the cursor is. 
+5. The search will automatically go around when it reaches the starting point. 
+
+### Accessibility
+
+This feature could help multitaksing terminal users. 

```

Comment:
Thankfully, we should benefit from the UIA provided directly from XAML. But we should consider how it's added to the UIA/XAML tree? Is the find box attached to the TermControl or to the TermApp?

As mentioned earlier, how can a user access all of this functionality without the use of a mouse? Are keybindings customizeable?

Our UIA providers need to be hooked up to this search functionality as well. That one isn't too difficult. Just a matter of putting what we need in the "FindText" function. I forget if it's in TermControlUiaProvider or UiaTextRange (I'd assume the first).

Reply:
Super curious what this would look like from a screen reader perspective as well. When the text is found, what does Narrator say? @carlos-zamora  your thoughts and experience w/ conhost would be helpful here.

Reply:
My current design is that SearchBox is attached to TermControl.

Reply:
Selection is tested with Accessibility Insights, and it could correctly find the selection.